cmd/thanos/query.go

```diff
@@ -422,94 +472,89 @@ func runQuery(
 			dialOpts,
```

Comment:
Am I wrong if I would say that if we somehow pass through spec the dial options per addresses we could have just one storeset right? Do you think it would be worth doing?

Reply:
Thanks for the review.
> if we somehow pass through spec the dial options per addresses we could have just one storeset right?
 
But dialopts are different for each endpoint config (as it depends on `config.tlsConfig`). So even if we somehow pass it through spec, we would not have one storeset as storeset would be defined inside the loop to have all tlsConfig.

To have only one storeset I thought to take the storeset initialization outside the loop but we can't because it uses dnsProvider which needs to be different for each endpoint config else it gives error.

So we need to have storeset inside the loop (which means more than one storeset) to have different dnsProvider and tlsConfig corresponding to different endpoint sets.

> Do you think it would be worth doing?

As discussed with @kakkoyun (in our 1:1) we expect only few configs (like 4 or 5) from the user so having these much of storesets would not affect the performance much. :)


<!--
Ah so, for one storeset we need to define it once, outside the for loop right?
I thought of your idea as taking the [storeset](https://github.com/thanos-io/thanos/pull/4389/files#diff-f9808a0ac67a474a18f6a7ea767593f85c216aeafe322c1ff5d17c81ababff95R427) outside the loop and then passing dialopts through spec by iterating over endpoint config but found that it is used at many places like store.update() which is inside the loop.
--!>

Reply:
> But dialopts are different for each endpoint config (as it depends on config.tlsConfig). So even if we somehow pass it through spec, we would not have one storeset as storeset would be defined inside the loop to have all tlsConfig.

Yea, I think we could do that per addresses.

> As discussed with @kakkoyun (in our 1:1) we expect only few configs (like 4 or 5) from the user so having these much of storesets would not affect the performance much. :)

I know companies running Thanos against 300 of StoreAPIs, so I would be careful here. But also I am not worried about performance too much here, rather maintainability and debuggability when you have different metrics per 100 of those storesets potentially. And different health loops. Not a biggie, maybe I can try propose a PR to yours if we want to compare different solutions? (:

Reply:
> I can try propose a PR to yours if we want to compare different solutions

That would be really helpful 🤗